feat: add enum variants

ChristopherRabotin opened the pull request:
This PR adds enum variants to simple types. The tests are separate from the enum function in serde.rs because these require a type annotation. 

This is a zeroth implementation of #242 . A subsequent PR, if I can figure it out (likely with help), will enable recursive types.

This PR also runs clippy on the codebase, cleaning up some legacy implementation. This is a separate commit to make it easy to undo.

ChristopherRabotin committed:
feat: add enum variants

ChristopherRabotin committed:
Run clippy

Nadrieril commented:
Could you revert the `Cargo.lock` version update? dhall-rust supports rust 1.60 (I don't mind bumping if there's a good reason, but I don't see one right now). EDIT: did it myself

Nadrieril commented on serde_dhall/tests/enum_tests.rs:
```diff
@@ -0,0 +1,92 @@
+        Unitary,
+    }
+
+    fn build_type() -> SimpleType {
+        let ty: SimpleType = serde_dhall::from_str(
+            r#"
+let ParentStruct = {id0: Integer, id1: Integer }
+
+let EnumVariant =
+        < SimpleStruct : { x : Double, y : Double, z : Double }
+        | InheritStruct: { field_a: ParentStruct, field_b: ParentStruct }
+        | Unitary
+        >       
+in EnumVariant"#,
+        )
+        .parse()
+        .unwrap();
+        ty
+    }
```
Doesn't the `StaticType` derive give you this type for free?

ChristopherRabotin replied:
It does! Committing the fix now.

Nadrieril commented on Cargo.lock:
```diff
@@ -1,6 +1,6 @@
 # This file is automatically @generated by Cargo.
 # It is not intended for manual editing.
-version = 3
+version = 4
```
```suggestion
version = 3
```

Nadrieril committed:
Revert `Cargo.lock` version change

Nadrieril commented:
Ah, it seems many of my dependencies have raised their MSRV (both Min and Max version it seems, even) :')

ChristopherRabotin commented:
Indeed, I'm seeing that the wasm crate requires a newer version of Rust: I'll check what's the MSRV I can get away with.

I'll check if StaticType supports the reflection, the manual building might be a relic of when I was trying the recursion.

ChristopherRabotin committed:
!fix: MSRV 1.76 for wasm_bindgen

Also remove unnecessary simple type definition

ChristopherRabotin commented:
WASM Bindgen MSRV is 1.76, which was released in [February 2024](https://blog.rust-lang.org/2024/02/08/Rust-1.76.0/), about fifteen months ago now.

ChristopherRabotin committed:
fix: versions in documentation

ChristopherRabotin commented:
I thought that the tests passed locally the other day, but now they don't anymore ... and the errors are not related to the changes I'm proposing here. (Edit: It's likely that I only tested the `enum` tests actually)

When running the tests are the root of the crate:
```
thread 'syntax::text::parser::test_grammar_files_in_sync' panicked at dhall/src/syntax/text/parser.rs:1056:9:
The local dhall.abnf file differs from the one from dhall-lang!
```

When running from the serde_dhall folder:
```
failures:

---- serde::test_file stdout ----

thread 'serde::test_file' panicked at serde_dhall/tests/serde.rs:290:9:
assertion `left == right` failed
  left: Err("No such file or directory (os error 2)")
 right: Ok(true)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- serde::test_import stdout ----

thread 'serde::test_import' panicked at serde_dhall/tests/serde.rs:12:9:
assertion `left == right` failed
  left: Err("Type error: error: error\n --> <current file>:1:1\n  |\n1 | ../dhall-lang/tests/parser/success/unit/BoolLitTrueA.dhall\n  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ No such file or directory (os error 2)\n  |")
 right: Ok(true)


failures:
    serde::test_file
    serde::test_import
```

Any idea how I could fix these? Or do these only work on Github for some reason?

I'll make another commit now that runs clippy on the other crates if it's OK. Since these changes are in their own commit, you should be able to trivially revert them if you prefer.

ChristopherRabotin committed:
Clippy changes

ChristopherRabotin committed:
fix: incorrect clippy fix in previous commit

Nadrieril commented:
Could you make a separate PR for the clippy fixes? Regarding the dhall.abnf error, did you maybe change the dhall-lang submodule to another commit than the one in version control?

Nadrieril commented:
Btw a MSRV change is not a minor change, we should bump to `0.13`. I've cherry-picked those commits to https://github.com/Nadrieril/dhall-rust/pull/245 so we can get things working again before we try to add any new feature.

ChristopherRabotin commented:
Sure thing! I've updated the version to 0.13.0 in the three different files that require the change, and undid the clippy changes. However, in updating the version, there will necessarily be a change in the Cargo.lock file.

```
diff --git a/Cargo.lock b/Cargo.lock
index 7b6794d..c341d3c 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1,6 +1,6 @@
 # This file is automatically @generated by Cargo.
 # It is not intended for manual editing.
-version = 3
+version = 4

 [[package]]
 name = "abnf"
@@ -1154,7 +1154,7 @@ dependencies = [

 [[package]]
 name = "serde_dhall"
-version = "0.12.2"
+version = "0.13.0"
 dependencies = [
  "dhall",
  "dhall_proc_macros",
lines 1-21/21 (END)
```

Concerning the submodule, I had not initialized it, so I don't think I had modified the commit it points to. I'm not entirely sure what is happening there.

_Edit:_ When fetching for your repo (instead of my fork), git reports that the referenced commit on dhall-lang does not exist:

```
Fetching submodule dhall-lang
fatal: remote error: upload-pack: not our ref bf9783fc4298d5d54897af1631d677b05dd19db5
Errors during submodule fetch:
	dhall-lang
```

ChristopherRabotin committed:
fix: version bump 0.13.0

ChristopherRabotin commented:
I'll merge #245 into this branch to avoid conflicts with the objective of merging 245 before this PR.

Nadrieril commented:
I did the changelog and version updates myself; could you rebase on top of master?

ChristopherRabotin committed:
Merge remote-tracking branch 'official/update' into feat/enum-variant-simple-type-issue-242-impl0

Nadrieril commented:
Regarding dhall I'm on 204a9d9dd167d2c9038539148a09825ded62f1b8, how did you get that other commit?

ChristopherRabotin committed:
Merge remote-tracking branch 'official/master' into feat/enum-variant-simple-type-issue-242-impl0

ChristopherRabotin committed:
doc: add enum struct variant to changelog

ChristopherRabotin commented:
Sorry, I got pretty confused trying to undo the clippy changes, and with the various merges. So I cherry picked the commit that implements the enum struct variants onto its own branch here: https://github.com/Nadrieril/dhall-rust/pull/246 .
